File: Algorithms/198. House Robber/golang/solution.recursion.go
func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	var memo map[int]int = make(map[int]int, len(nums))
	return robFrom(memo, 0, nums)
}

func robFrom(memo map[int]int, i int, nums []int) int {
	if i >= len(nums) {
		return 0
	}

	if _, ok := memo[i]; ok {
		return memo[i]
	}
	if robFrom(memo, i+1, nums) > robFrom(memo, i+2, nums)+nums[i] {
		memo[i] = robFrom(memo, i+1, nums)
	} else {
		memo[i] = robFrom(memo, i+2, nums) + nums[i]
	}
	return memo[i]
}

File: Algorithms/198. House Robber/php/solution.recursion.php
<?php
class Solution
{
    private $memo = [];

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function rob($nums)
    {
        return $this->robFrom(0, $nums);
    }

    function robFrom($i, $nums)
    {
        if ($i >= count($nums)) {
            return 0;
        }
        if (isset($this->memo[$i])) {
            return $this->memo[$i];
        }
        $ans = max($this->robFrom($i + 1, $nums), $this->robFrom($i + 2, $nums) + $nums[$i]);
        $this->memo[$i] = $ans;
        return $ans;
    }
}
